frame/curve-amm/src/maths.rs

```diff
@@ -97,30 +100,29 @@ pub fn compute_base(new_quote: u128, amp_coeff: u128, d: u128) -> Result<u128, D
 	let one = to_big_uint(1_u128);
 	let mut d = to_big_uint(d);
 	let amplification_coefficient = to_big_uint(amp_coeff);
-	let mut ann = amplification_coefficient.mul(&n).mul(&n);
+	let ann = amplification_coefficient.mul(&n).mul(&n);
 
 	// s and p are same as input base amount as pool supports only 2 assets.
 	let s = to_big_uint(new_quote);
 	let mut p = to_big_uint(new_quote);
-	// b = s + (d / ann) -d
-	// c = d^(n + 1) / (ann * n^n * p)
 
-	let d_ann = safe_div(&mut d, &mut ann)?;
+	// term1 = d^(n + 1) / n^n * p
+	// term2 = 2*y + s - d
+
 	let d_n = safe_div(&mut d, &mut n)?;
-	let b = s.add(&d_ann); // substract d later
-	let mut c = d_ann.mul(&d_n).mul(&d_n);
-	let c = safe_div(&mut c, &mut p)?;
+	let mut c = d_n.clone().mul(&d_n).mul(&d);
+	let term1 = safe_div(&mut c, &mut p)?;
 
 	let mut y = d.clone();
 
-	// y = (y^2 + c) / (2y + b)
+	// y = (y^2 * ann + term1) / (ann * term2 + d)
```

Comment:
as per paper, can calculate terms once, and use in iteration. as terms are constants. why not here?

Reply:
In paper there is division by `ann` which prevents `ann` to be 0. So I have rewrite formulas so that things work fine even if `ann` is 0. As per that `term1` can be reused but `term2` requires `y` so it must be recomputed. 
See comments above.

Reply:
is model valid is ann = 0? or it degenrates to some constant product? does viper also handles ann zero as we do? could it overflow if ann is 0.0000000000000000001 (close to zero)?

Reply:
As per paper if `amp` is 0 it is equal to constant product. I think viper does not handle this. 
for last question I don't have very confident answer for following equation but to have underflow it needs to satisfy following
`(n + 1) * d_p < d`
it can be simplified for `n = 2` and by substituting equations
` 3 * (x + dx)^2 + 2 (x + dx) (y + dy) + 3 * (y + dy)^2  < 0`
x and y current pool balances ,  dx = changes in X and dy = changes in y.
So it seems it can only be negative if either dx or dy is negative but I think that is not possible.
I have also added proptest for pool with amp = 0. and they passes.
Please let me know if there is issues.

Reply:
non handling zero is feature imho, not a bug

Reply:
sure I agree.